runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformExecutor.java

```diff
@@ -104,19 +104,23 @@ public void run() {
         return;
       }
 
-      processElements(evaluator, metricsContainer, enforcements);
+      processElements(evaluator, enforcements);
 
-      finishBundle(evaluator, metricsContainer, enforcements);
+      finishBundle(evaluator, enforcements);
+
+      synchronized (context) {
+        context.getCommittedMetrics().update(stepName, metricsContainer);
+      }
     } catch (Exception e) {
       onComplete.handleException(inputBundle, e);
       if (e instanceof RuntimeException) {
         throw (RuntimeException) e;
       }
       throw new RuntimeException(e);
     } finally {
-      // Report the physical metrics from the end of this step.
-      context.getMetrics().commitPhysical(inputBundle, metricsContainer.getCumulative());
-
+      synchronized (context) {
```

Comment:
The reason for synchronization is exactly what you remarked on, the evaluation context is shared between several threads.

There is synchronization in the existing metrics implementation for this reason as well [DirectMetrics.java](https://github.com/apache/beam/blob/v0.6.0/runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java#L114).

I agree that each thread having its own evaluation context would be better but I'm not sure this should be changed as part of this PR.

I will see if there is a different point in the flow in which the containers can be merged without synchronization.